File: leetcode/0720-longest-word-in-dictionary/Solution.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    public String longestWord(String[] words) {
        Set<String> wordSet = new HashSet<>();
        for (String w : words) wordSet.add(w);

        String best = "";
        for (String word : words) {
            boolean buildable = true;
            for (int i = 1; i <= word.length(); i++) {
                if (!wordSet.contains(word.substring(0, i))) {
                    buildable = false;
                    break;
                }
            }
            if (buildable) {
                if (word.length() > best.length() || (word.length() == best.length() && word.compareTo(best) < 0)) {
                    best = word;
                }
            }
        }
        return best;
    }
}

File: leetcode/0720-longest-word-in-dictionary/solution.cpp
#include <vector>
#include <string>
#include <unordered_set>
using namespace std;

class Solution {
public:
    string longestWord(vector<string>& words) {
        unordered_set<string> wordSet(words.begin(), words.end());

        string best = "";
        for (const string& word : words) {
            bool buildable = true;
            for (size_t i = 1; i <= word.size(); i++) {
                if (wordSet.find(word.substr(0, i)) == wordSet.end()) {
                    buildable = false;
                    break;
                }
            }
            if (buildable) {
                if (word.size() > best.size() || (word.size() == best.size() && word < best)) {
                    best = word;
                }
            }
        }
        return best;
    }
};
